### DISCUSSION TITLE
Centering window on a secondary monitor

### DISCUSSION THREAD

**Discussion by fcoury** [Hammerspoon/hammerspoon#3302] [General]
I am not sure if I'm doing something wrong but I am trying to center the window and it works fine on the main monitor but doesn't work for any other monitors.

Here's there relevant code:

```lua
hs.hotkey.bind({"cmd", "alt", "ctrl"}, "V", function()
  local win = hs.window.focusedWindow()
  local f = win:frame()
  local screen = win:screen()
  local max = screen:frame()

  local w = max.w*0.6
  local h = max.h*0.6

  f.x = max.w/2 - w/2
  f.y = max.h/2 - h/2
  f.w = w
  f.h = h
  win:setFrame(f, 0)
end)
```

Can anyone point me in the right direction on how I can make this work accross different monitors?

Thanks!

**Comment by fcoury** [comment]
Put something together now but not sure how correct this is:

```lua
function resizeAndCenterWinByPerc(perc)
  local win = hs.window.focusedWindow()
  local f = win:frame()
  local screen = win:screen()
  local max = screen:frame()

  local w = max.w*perc
  local h = max.h*perc

  f.w = w
  f.h = h

  win:setFrame(f, 0)

  f.x = max.w/2 - w/2
  f.y = max.h/2 - h/2

  nf = screen:localToAbsolute(f)
  win:move(nf, screen, true, 0)
end

hs.hotkey.bind({"cmd", "alt", "ctrl"}, "C", function()
  resizeAndCenterWinByPerc(0.8)
end)

hs.hotkey.bind({"cmd", "alt", "ctrl"}, "V", function()
  resizeAndCenterWinByPerc(0.6)
end)
```

**Reply by muescha** [reply]
to compare: I have done it this way. i don't center it. but you can see how I have done it. here hyper+2 put the window on the lower external screen:

```lua
hs.hotkey.bind(hyper, "2", keyInfo("place on other half of monitor"), function()
    local externalScreen = hs.screen.allScreens()[2]
    local window = hs.window.frontmostWindow()
    window:moveToScreen(externalScreen, false, true)
    if externalScreen:frame().w > externalScreen:frame().h then
        window:setFrame({
            x = externalScreen:frame().x + (externalScreen:frame().w / 2),
            y = externalScreen:frame().y,
            w = externalScreen:frame().w / 2,
            h = externalScreen:frame().h,
        })
    else
        window:setFrame({
            x = externalScreen:frame().x,
            y = externalScreen:frame().y + (externalScreen:frame().h / 2),
            w = externalScreen:frame().w,
            h = externalScreen:frame().h / 2,
        })
    end
end)
```

https://github.com/muescha/dot_hammerspoon/blob/fd851f32075549ee5a2c7da4a0488f0e753f0918/Functions/WindowPlacer.lua#L47

**Comment by Bellavene** [comment]
macos has this feature built-in... fn+ctrl+c
Menu > Window > Center

**Comment by dmgerman** [comment]
I think the easiest solution is to think of it as two different operations, each with its own function:

- Move window to desired screen
- Resize in the current display

That will handle any number of external monitors. If an operation is common, create a small function that calls these two functions one after the other.

      hs.window.focusedWindow():screen()

will return the screen where the window is.
